### DISCUSSION TITLE
SwitchMode 1 working incorrectly if toggle relay using Web GUI

### DISCUSSION THREAD

**Discussion by zakdma** [arendst/Tasmota#22574] [General]
Hey!

I have a question that I can't resolve with SwitchMode setting. Maybe I don't understand it correctly and please point me in the right direction then.
I have Wemos D1 mini and Tasmota 14.3.0 installed on it.
I configured Module type to Generic (18).
Also I configured a switch connected to GPIO14 (and GND) and a relay connected to GPIO13.
I run command ```SwitchMode1 1``` which means (as I understood) that relay WILL FOLLOW the switch. So if the switch is open then the relay is OFF, and if the switch is closed (connected to GND) then relay is ON.
The current state after power up is OFF. All is working fine as described.
BUT then I go to Tasmota Web GUI and click Toggle button https://prnt.sc/rRk6oUmGLrXx
It makes the relay ON but the switch is still OPEN. And then, when I close the switch it makes the relay OFF instead of ON. So relay is inverted and does NOT FOLLOW the switch status anymore. It's very strange behavior and looks like a bug.
Could you please explain what is happening, and how to workaround it?

Thanks!

**Comment by sfromis** [comment]
Using `SwitchMode` is irrelevant to the web UI toggles, as these are not simulating the Switch pin *input*, instead acting on the Relay pin *output*, in a similar way as the `Power` command.

**Reply by zakdma** [reply]
Could you please explain it's correct that ```Switch``` should short GND to pin?
Because I see that open ```Switch``` shows ```ON``` in a ```status 10``` command. And closed to GND ```Switch``` shows ```OFF```.
Is it correct behavior?
Also what about MQTT? If Tasmota received MQTT to activate the relay behavior will be sthe ame as I do it using web UI?

**Reply by sfromis** [reply]
It is a common standard that a Switch (or Button) input will connect the pin to GND when activated, with internal pull-up enabled in the MCU. It is quite normal that electrical low gives "OFF" in `Status 10` output, and electrical high (the internal pull-up) gives "ON". That's arguably consistent at the physical level.

However, do not get confused mixing up terms like ON/OFF/1/0/high/low etc, there is no inherent meaning for that when it comes to what the desired behavior is, but you have to match your config to get from the physical to the behavior fitting your use case. Still, the starting point for Tasmota is the standard with GND and internal pull-up, with options to config for different types of circuit design.

Actions to change output Relay state are indeed the same, no matter if they come via MQTT, console command, web UI toggle, rules, or are the result of a Switch/Button action. And remember that a Switch/Button input is not remembering any state, they are instead about acting on transitions between high and low.

**Reply by zakdma** [reply]
> It is a common standard that a Switch (or Button) input will connect the pin to GND when activated, with internal pull-up enabled in the MCU. It is quite normal that electrical low gives "OFF" in `Status 10` output, and electrical high (the internal pull-up) gives "ON". That's arguably consistent at the physical level.
> 
> However, do not get confused mixing up terms like ON/OFF/1/0/high/low etc, there is no inherent meaning for that when it comes to what the desired behavior is, but you have to match your config to get from the physical to the behavior fitting your use case. Still, the starting point for Tasmota is the standard with GND and internal pull-up, with options to config for different types of circuit design.
> 
> Actions to change output Relay state are indeed the same, no matter if they come via MQTT, console command, web UI toggle, rules, or are the result of a Switch/Button action. And remember that a Switch/Button input is not remembering any state, they are instead about acting on transitions between high and low.

Sorry, but I don't understand this logic. I started from scratch. I suppose default values should work as expected. Open switch then Relay is OFF. Closed switch then Relay is ON.
So I configured Switch1 on GPIO14 and Relay1 on GPIO13 https://prnt.sc/Vne1EX0WE8Jf
After reset, I see this in console https://prnt.sc/FCmSCRn-n9ox
- "Switch1": "ON"
- "POWER": "OFF"
Physical pins:
- GPIO14 (Switch) is HIGH
- GPIO13 (Relay) is LOW. (Relay is deactivated)
Ok, toggle switch to ON (close it to GND)
In console I see this https://prnt.sc/f6XZUZby7xW_
- "POWER": "OFF"
- "Switch1": "OFF"
Relay is still inactive (surprise!!!)
Physical pins:
- GPIO14 (Switch) is LOW
- GPIO13 (Relay) is LOW. (Relay is still deactivated)
Ok, toggle switch to OFF (open it) 
In console I see this https://prnt.sc/oEPhVtTUwR6q
- "POWER": "ON"
- "Switch1": "ON"
Relay is activated (surprise!!!)
Physical pins:
- GPIO14 (Switch) is HIGH
- GPIO13 (Relay) is HIGH.

Ok, in general it follows the switch but in inverse order. Why? Why default behaviour is not switch is ON and Relay is ON?
Locally (inside one module) I can inverse relay control so I will have the desired behavior.
But that caused me problems with device group. Imagine I want to control remote relay. So I configured relay to use DeviceGroup the same as module with switch. So again Switch is open relay (surprise!!!) is ON. I don't want to inverse this remote relay because it is controlled also by a HA by MQTT. How can I achieve this without using Switch_n and creating a circuit with a pull-down resistor?

Thank you!

**Reply by sfromis** [reply]
The logic is that when the Switch input is electrically low, pulled down by the contact, it will show as "OFF". And when engaged, it will be electrically high (due to the pullup), showing as "ON" in `Status 10` output. What you'd see if having a LED connected there. Such things are a reason why I reminded you to not get confused about those terms. This *intentional* way of working is quite normal, and I'd even recommend to not worry too much about what you see in `Status 10` and `State` output. Just get used to it :wink: 

Sure, selecting an inverted `SwitchMode` would invert this. While this on a superficial level would look sensible, it is just not how it works when engaging the input pulls it **down**, therefore becoming "OFF" when engaged. You should not try to "fix" this by inverting the relay do be the opposite. Of course, many relays do have both NO/NC screws, normally open/closed, just to confuse things even more....

Only start to combine with other ways of triggering the relay after having the basics set up correctly.

**Reply by zakdma** [reply]
@sfromis thank you for help.
I resolved the problem by using switchmode 2 instead of 1. It is inverted and works as I expected.

**Reply by sfromis** [reply]
Whatever works for you....

Still that's technically the wrong SwitchMode for a regular input, and will make the Relay behave wrongly, unless you also make the "wrong" config of inverted there - which would not work well with intentional actions to switch on/off the relay.
